src/main/java/org/junitpioneer/jupiter/CartesianProductTestExtension.java

```diff
@@ -94,17 +102,50 @@ private static void ensureNoInputConflicts(CartesianProductTest annotation,
 		return sets;
 	}
 
-	private List<List<?>> getSetsFromRepeatableAnnotation(List<CartesianValueSource> valueSources) {
+	private List<List<?>> getSetsFromArgumentsSources(List<? extends Annotation> argumentsSources,
+			ExtensionContext context) {
 		List<List<?>> sets = new ArrayList<>();
-		for (CartesianValueSource source : valueSources) {
-			CartesianValueArgumentsProvider provider = new CartesianValueArgumentsProvider();
-			provider.accept(source);
-			List<Object> collect = provider.provideArguments().distinct().collect(toList());
-			sets.add(collect);
+		for (Annotation source : argumentsSources) {
+			sets.add(getSetFromAnnotation(context, source));
 		}
 		return sets;
 	}
 
+	private List<Object> getSetFromAnnotation(ExtensionContext context, Annotation source) {
+		try {
+			ArgumentsProvider provider = initializeArgumentsProvider(source);
+			return provideArguments(context, source, provider);
+		}
+		catch (Exception e) {
+			throw new ExtensionConfigurationException("Could not provide arguments because of exception.", e);
+		}
```

Comment:
```suggestion
		catch (Exception ex) {
			throw new ExtensionConfigurationException("Could not provide arguments because of exception.", ex);
		}
```